diff --git a/include/boost/dynamic_bitset/dynamic_bitset.hpp b/include/boost/dynamic_bitset/dynamic_bitset.hpp
index a2abfc9..ba0e8ef 100644
--- a/include/boost/dynamic_bitset/dynamic_bitset.hpp
+++ b/include/boost/dynamic_bitset/dynamic_bitset.hpp
@@ -712,54 +712,54 @@ swap(dynamic_bitset<Block, Allocator>& b) // no throw
     std::swap(m_num_bits, b.m_num_bits);
 }
 
 template <typename Block, typename Allocator>
 dynamic_bitset<Block, Allocator>& dynamic_bitset<Block, Allocator>::
 operator=(const dynamic_bitset<Block, Allocator>& b)
 {
     m_bits = b.m_bits;
     m_num_bits = b.m_num_bits;
     return *this;
 }
 
 #ifndef BOOST_NO_CXX11_RVALUE_REFERENCES
 
 template <typename Block, typename Allocator>
 inline dynamic_bitset<Block, Allocator>::
 dynamic_bitset(dynamic_bitset<Block, Allocator>&& b)
   : m_bits(boost::move(b.m_bits)), m_num_bits(boost::move(b.m_num_bits))
 {
     // Required so that assert(m_check_invariants()); works.
-    assert((b.m_bits = buffer_type()).empty());
+    assert((b.m_bits = buffer_type(get_allocator())).empty());
     b.m_num_bits = 0;
 }
 
 template <typename Block, typename Allocator>
 inline dynamic_bitset<Block, Allocator>& dynamic_bitset<Block, Allocator>::
 operator=(dynamic_bitset<Block, Allocator>&& b)
 {
     if (boost::addressof(b) == this) { return *this; }
 
     m_bits = boost::move(b.m_bits);
     m_num_bits = boost::move(b.m_num_bits);
     // Required so that assert(m_check_invariants()); works.
-    assert((b.m_bits = buffer_type()).empty());
+    assert((b.m_bits = buffer_type(get_allocator())).empty());
     b.m_num_bits = 0;
     return *this;
 }
 
 #endif // BOOST_NO_CXX11_RVALUE_REFERENCES
 
 template <typename Block, typename Allocator>
 inline typename dynamic_bitset<Block, Allocator>::allocator_type
 dynamic_bitset<Block, Allocator>::get_allocator() const
 {
     return m_bits.get_allocator();
 }
 
 //-----------------------------------------------------------------------------
 // size changing operations
 
 template <typename Block, typename Allocator>
 void dynamic_bitset<Block, Allocator>::
 resize(size_type num_bits, bool value) // strong guarantee
 {
