--- /usr/lib/python3/dist-packages/mesonbuild/compilers/mixins/clike.py	2025-03-20 22:15:41.867561135 -0700
+++ /usr/lib/python3/dist-packages/mesonbuild/compilers/mixins/clike.py.bak	2025-03-20 22:16:03.614515077 -0700
@@ -75,19 +75,11 @@
         # TODO: this could probably be added to the DynamicLinker instead
         if isinstance(self.compiler.linker, (GnuLikeDynamicLinkerMixin, SolarisDynamicLinker, CompCertDynamicLinker)):
             group_start = -1
             group_end = -1
-            ignore = False
             for i, each in enumerate(new):
-                if ignore:
-                    ignore = False
-                    continue
-                if each == "-Xclang":
-                    ignore = True
-                    continue
                 if not GROUP_FLAGS.search(each):
                     continue
-
                 group_end = i
                 if group_start < 0:
                     # First occurrence of a library
                     group_start = i
