diff --git a/dlls/ntdll/unix/debug.c b/dlls/ntdll/unix/debug.c
index e84a47f58a4..06501a502a3 100644
--- a/dlls/ntdll/unix/debug.c
+++ b/dlls/ntdll/unix/debug.c
@@ -34,4 +34,5 @@
 #include <sys/stat.h>
 #include <unistd.h>
+#include <errno.h>
 
 #include "ntstatus.h"
@@ -60,4 +61,5 @@ static BOOL init_done;
 static struct debug_info initial_info;  /* debug info for initial thread */
 static unsigned char default_flags = (1 << __WINE_DBCL_ERR) | (1 << __WINE_DBCL_FIXME);
+int logfd = 2;
 static int nb_debug_options = -1;
 static int options_size;
@@ -201,4 +203,5 @@ static void debug_usage(void)
 static void init_options(void)
 {
+    char *wine_debug_log = getenv("WINEDEBUGLOG");
     char *wine_debug = getenv("WINEDEBUG");
     const char *app_name, *p;
@@ -207,4 +210,18 @@ static void init_options(void)
     nb_debug_options = 0;
 
+    if (wine_debug_log) {
+        logfd = open(wine_debug_log, O_CREAT | O_WRONLY | /* O_TRUNC */ O_APPEND, 0666);
+        if (logfd < 0) {
+            char buffer[MAX_PATH + 1];
+
+            int slen = snprintf(buffer, sizeof(buffer),
+                                "could open wine debug log at \"%s\": %s",
+                                wine_debug_log, strerror(errno));
+
+            write(2, buffer, slen);
+            exit(1);
+        }
+    }
+
     /* check for stderr pointing to /dev/null */
     if (!fstat( 2, &st1 ) && S_ISCHR(st1.st_mode) &&
@@ -280,5 +297,5 @@ NTSTATUS unixcall_wine_dbg_write( void *args )
     struct wine_dbg_write_params *params = args;
 
-    return write( 2, params->str, params->len );
+    return write( logfd, params->str, params->len );
 }
 
@@ -311,5 +328,5 @@ int __cdecl __wine_dbg_output( const char *str )
     {
         ret += append_output( info, str, end + 1 - str );
-        write( 2, info->output, info->out_pos );
+        write( logfd, info->output, info->out_pos );
         info->out_pos = 0;
         str = end + 1;
diff --git a/dlls/ntdll/unix/virtual.c b/dlls/ntdll/unix/virtual.c
index ca5d50b0fe6..fe2679d8f76 100644
--- a/dlls/ntdll/unix/virtual.c
+++ b/dlls/ntdll/unix/virtual.c
@@ -84,4 +84,6 @@
 #include "wine/debug.h"
 
+extern int logfd;
+
 WINE_DEFAULT_DEBUG_CHANNEL(virtual);
 WINE_DECLARE_DEBUG_CHANNEL(module);
@@ -225,5 +227,5 @@ static const size_t view_block_size = 0x100000;
 static void *preload_reserve_start;
 static void *preload_reserve_end;
-static BOOL force_exec_prot;  /* whether to force PROT_EXEC on all PROT_READ mmaps */
+static BOOL force_exec_prot = TRUE;  /* whether to force PROT_EXEC on all PROT_READ mmaps */
 static BOOL enable_write_exceptions;  /* raise exception on writes to executable memory */
 
@@ -2844,4 +2846,19 @@ static IMAGE_BASE_RELOCATION *process_relocation_block( char *page, IMAGE_BASE_R
 }
 
+static ssize_t path_of_fd(int fd, char *out, size_t len) {
+  ssize_t ret;
+
+  char buff[256];
+  snprintf(buff, sizeof(buff), "/proc/self/fd/%d", fd);
+
+  ret = readlink(buff, out, len);
+
+  if (ret > 0 && ret < len)
+    out[ret] = '\0';
+  else
+    return -1;
+
+  return ret;
+}
 
 /***********************************************************************
@@ -2865,4 +2882,5 @@ static NTSTATUS map_image_into_view( struct file_view *view, const WCHAR *filena
     char *header_end;
     char *ptr = view->base;
+    char fd_path[MAX_PATH+1];
     SIZE_T header_size, header_map_size, total_size = view->size;
     SIZE_T align_mask = max( image_info->alignment - 1, page_mask );
@@ -2983,4 +3001,5 @@ static NTSTATUS map_image_into_view( struct file_view *view, const WCHAR *filena
         TRACE_(module)( "mapping %s section %.8s at %p off %x size %x virt %x flags %x\n",
                         debugstr_w(filename), sec[i].Name, ptr + sec[i].VirtualAddress,
+                        logfd != 2 && path_of_fd(fd, fd_path, sizeof(fd_path)) > 0 ? fd_path : debugstr_w(filename), sec[i].Name, ptr + sec[i].VirtualAddress,
                         (int)sec[i].PointerToRawData, (int)sec[i].SizeOfRawData,
                         (int)sec[i].Misc.VirtualSize, (int)sec[i].Characteristics );
